File: go/parser.go
package main

import (
	"errors"
	"regexp"
	"strconv"
	"strings"
	"time"
)

var InvalidCommand error = errors.New("Invalid command")

const (
	Get = iota
	Set
	Del
	GetCounter
	SetCounter
	DelCounter
	GetDump
	NewDump
	DumpInterval
	SetTTL
	Upload
	Download
	Remove
)

const GetString = "GET"
const SetString = "SET"
const DelString = "DEL"
const GetCounterString = "GETC"
const SetCounterString = "SETC"
const DelCounterString = "DELC"
const GetDumpString = "GETDUMP"
const NewDumpString = "NEWDUMP"
const DumpIntervalString = "DUMPINTERVAL"
const SetTTLString = "SETTTL"
const UploadString = "UPLOAD"
const DownloadString = "DOWNLOAD"
const RemoveString = "REMOVE"

const regExp = "[[:alnum:]]+"

var matcher, _ = regexp.Compile(regExp)

const timeExp = "([0-9][0-9])h-([0-9][0-9])m-([0-9][0-9])s"

var timeMatcher, _ = regexp.Compile(timeExp)

func parseDuration(input string) *time.Duration {
	subMatches := timeMatcher.FindStringSubmatch(input)
	if len(subMatches) != 4 {
		return nil
	}
	var hours int64
	var minutes int64
	var seconds int64

	hours, err := strconv.ParseInt(subMatches[1], 10, 32)
	if err != nil {
		return nil
	}
	minutes, err = strconv.ParseInt(subMatches[2], 10, 32)
	if err != nil {
		return nil
	}
	seconds, err = strconv.ParseInt(subMatches[3], 10, 32)
	if err != nil {
		return nil
	}
	minutes += hours * 60
	seconds += minutes * 60
	ret := time.Second * time.Duration(seconds)
	return &ret
}

type CompleteCommand struct {
	CommandKind uint16
	Key         string
	Value       string
	Size        int64
	Ttl         *time.Duration
}

func InterpretCommand(command string) (*CompleteCommand, error) {

	parts := strings.Split(command, " ")
	if len(parts) == 0 {
		return nil, InvalidCommand
	}
	commandString := parts[0]
	retCommand := CompleteCommand{}

	switch commandString {
	case GetString:
		if len(parts) != 2 {
			return nil, InvalidCommand
		}
		retCommand.CommandKind = Get
		potKey := parts[1]
		potKeyReg := matcher.MatchString(potKey)
		if !potKeyReg {
			return nil, InvalidCommand
		}
		retCommand.Key = potKey
	case SetString:
		if len(parts) != 3 {
			return nil, InvalidCommand
		}
		retCommand.CommandKind = Set
		potKey := parts[1]
		potVal := parts[2]
		potKeyReg := matcher.MatchString(potKey)
		potValReg := matcher.MatchString(potVal)
		if !(potKeyReg && potValReg) {
			return nil, InvalidCommand
		}
		retCommand.Key = potKey
		retCommand.Value = potVal
	case DelString:
		if len(parts) != 2 {
			return nil, InvalidCommand
		}
		retCommand.CommandKind = Del
		potKey := parts[1]
		potKeyReg := matcher.MatchString(potKey)
		if !potKeyReg {
			return nil, InvalidCommand
		}
		retCommand.Key = potKey
	case GetCounterString:
		if len(parts) != 1 {
			return nil, InvalidCommand
		}
		retCommand.CommandKind = GetCounter
	case SetCounterString:
		if len(parts) != 1 {
			return nil, InvalidCommand
		}
		retCommand.CommandKind = SetCounter
	case DelCounterString:
		if len(parts) != 1 {
			return nil, InvalidCommand
		}
		retCommand.CommandKind = DelCounter
	case GetDumpString:
		if len(parts) != 1 {
			return nil, InvalidCommand
		}
		retCommand.CommandKind = GetDump
	case NewDumpString:
		if len(parts) != 1 {
			return nil, InvalidCommand
		}
		retCommand.CommandKind = NewDump
	case DumpIntervalString:
		if len(parts) != 2 {
			return nil, InvalidCommand
		}
		duration := parseDuration(parts[1])
		if duration == nil {
			return nil, InvalidCommand
		}
		retCommand.CommandKind = DumpInterval
		retCommand.Ttl = duration
	case SetTTLString:
		if len(parts) != 4 {
			return nil, InvalidCommand
		}
		retCommand.CommandKind = SetTTL
		potKey := parts[1]
		potVal := parts[2]
		potKeyReg := matcher.MatchString(potKey)
		potValReg := matcher.MatchString(potVal)
		if !(potKeyReg && potValReg) {
			return nil, InvalidCommand
		}
		potDur := parts[3]
		duration := parseDuration(potDur)
		if duration == nil {
			return nil, InvalidCommand
		}
		retCommand.Key = potKey
		retCommand.Value = potVal
		retCommand.Ttl = duration
	case UploadString:
		if len(parts) != 3 {

			return nil, InvalidCommand
		}
		retCommand.CommandKind = Upload
		potKey := parts[1]
		potKeyReg := matcher.MatchString(potKey)
		if !potKeyReg {
			return nil, InvalidCommand
		}
		potSize := parts[2]
		size, err := strconv.ParseInt(potSize, 10, 64)
		if err != nil {
			return nil, InvalidCommand
		}
		retCommand.Size = size
		retCommand.Key = potKey
	case DownloadString:
		if len(parts) != 2 {
			return nil, InvalidCommand
		}
		retCommand.CommandKind = Download
		potKey := parts[1]
		potKeyReg := matcher.MatchString(potKey)
		if !potKeyReg {
			return nil, InvalidCommand
		}
		retCommand.Key = potKey
	case RemoveString:
		if len(parts) != 2 {
			return nil, InvalidCommand
		}
		retCommand.CommandKind = Remove
		potKey := parts[1]
		potKeyReg := matcher.MatchString(potKey)
		if !potKeyReg {
			return nil, InvalidCommand
		}
		retCommand.Key = potKey
	default:
		return nil, InvalidCommand
	}
	return &retCommand, nil
}

File: C/parser.c
//
// Created by Hannes Furmans on 01.04.22.
//
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <regex.h>

#include "parser.h"

regex_t regex;

void compile_regex() {
    regcomp(&regex, "[[:alnum:]]+", 0);
}

enum CommandType classify_command(char * token);

bool validate_string(char * token);

bool parse_time(char * string, struct tm * time_out) {
    strptime(string, "%Hh-%Mm-%Ss", time_out);
    if (time_out->tm_hour == 0 && time_out->tm_min == 0 && time_out->tm_sec == 0) {
        return false;
    }
    return true;
}

double convert_to_seconds(struct tm * time) {
    double ret = 0;
    ret += time->tm_sec;
    ret += time->tm_min * 60;
    ret += time->tm_hour * 60 * 60;
    return ret;
}

void remove_trailing(char * input) {
    input[strcspn(input, "\n")] = 0;
}

void invalidate_command(struct CompleteCommand * cmd) {
    cmd->type = Invalid;
}

struct CompleteCommand * parse(char * input) {
    struct CompleteCommand * ret = malloc(sizeof(struct CompleteCommand));

    remove_trailing(input);

    char * token = strtok(input, " ");

    unsigned char loop_count = 0;
    while (token != NULL) {
        if (loop_count == 0) {
            enum CommandType tmp_type = classify_command(token);
            if (tmp_type == Invalid) {
                invalidate_command(ret);
                return ret;
            }
            ret->type = tmp_type;
        } else if (loop_count == 1) {
            struct tm * parsed_time = malloc(sizeof(struct tm));
            bool parse_result = parse_time(token, parsed_time);
            if (!parse_result) {
                free(parsed_time);
                if (!validate_string(token)) {
                    invalidate_command(ret);
                    return ret;
                }
                ret->key = token;
            } else {
                ret->time = parsed_time;
            }
        } else if (loop_count == 2) {
            if (!validate_string(token)) {
                invalidate_command(ret);
                return ret;
            }
            ret->value = token;
            ret->value_len = strlen(token);
        } else if (loop_count == 3) {
            struct tm * parsed_time = malloc(sizeof(struct tm));
            bool time_result = parse_time(token, parsed_time);
            if (!time_result) {
                free(parsed_time);
                invalidate_command(ret);
                return ret;
            }
            ret->time = parsed_time;
        } else {
            invalidate_command(ret);
            return ret;
        }

        token = strtok(NULL, " ");
        loop_count++;
    }

    if (ret->type == Get && loop_count != 2) {
        invalidate_command(ret);
        return ret;
    } else if (ret->type == Set && loop_count != 3) {
        invalidate_command(ret);
        return ret;
    } else if (ret->type == Del && loop_count != 2) {
        invalidate_command(ret);
        return ret;
    } else if (ret->type == Get_Counter && loop_count != 1) {
        invalidate_command(ret);
        return ret;
    } else if (ret->type == Set_Counter && loop_count != 1) {
        invalidate_command(ret);
        return ret;
    } else if (ret->type == Del_Counter && loop_count != 1) {
        invalidate_command(ret);
        return ret;
    } else if (ret->type == Get_Dump && loop_count != 1) {
        invalidate_command(ret);
        return ret;
    } else if (ret->type == New_Dump && loop_count != 1) {
        invalidate_command(ret);
        return ret;
    } else if (ret->type == Dump_Interval && loop_count != 2) {
        invalidate_command(ret);
        return ret;
    } else if (ret->type == Set_TTL && loop_count != 4) {
        invalidate_command(ret);
        return ret;
    }

    return ret;
};

inline enum CommandType classify_command(char * token) {
    if (strcmp(token, GET_STRING) == 0)
        return Get;
    else if (strcmp(token, SET_STRING) == 0)
        return Set;
    else if (strcmp(token, DEL_STRING) == 0)
        return Del;
    else if (strcmp(token, GET_COUNTER_STRING) == 0)
        return Get_Counter;
    else if (strcmp(token, SET_COUNTER_STRING) == 0)
        return Set_Counter;
    else if (strcmp(token, DEL_COUNTER_STRING) == 0)
        return Del_Counter;
    else if (strcmp(token, GET_DUMP_STRING) == 0)
        return Get_Dump;
    else if (strcmp(token, NEW_DUMP_STRING) == 0)
        return New_Dump;
    else if (strcmp(token, DUMP_INTERVAL_STRING) == 0)
        return Dump_Interval;
    else if (strcmp(token, SET_TTL_STRING) == 0)
        return Set_TTL;
    else
        return Invalid;
}

inline bool validate_string(char * token) {
    int reti = regexec(&regex, token, 0, NULL, 0);
    if (reti)
        return true;
    else
        return false;
}

void complete_command_print(struct CompleteCommand * command) {
    if (command->type == Get) {
        printf("%s %s\n", GET_STRING, command->key);
    } else if (command->type == Set) {
        printf("%s %s %s\n", SET_STRING, command->key, command->value);
    } else if (command->type == Del) {
        printf("%s %s\n", DEL_STRING, command->key);
    } else if (command->type == Get_Counter) {
        printf("%s\n", GET_COUNTER_STRING);
    } else if (command->type == Set_Counter) {
        printf("%s\n", SET_COUNTER_STRING);
    } else if (command->type == Del_Counter) {
        printf("%s\n", DEL_COUNTER_STRING);
    } else if (command->type == Get_Dump) {
        printf("%s\n", GET_DUMP_STRING);
    } else if (command->type == New_Dump) {
        printf("%s\n", NEW_DUMP_STRING);
    } else if (command->type == Dump_Interval) {
        printf("%s %fs\n", DUMP_INTERVAL_STRING, convert_to_seconds(command->time));
    } else if (command->type == Set_TTL) {
        printf("%s %s %s %f\n", SET_TTL_STRING, command->key, command->value, convert_to_seconds(command->time));
    } else {
        printf("Command is invalid\n");
    }
};

